src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperRunner.cs

```diff
@@ -24,7 +24,10 @@ public class TagHelperRunner
                 executionContext.AllAttributes,
                 executionContext.UniqueId,
                 executionContext.GetChildContentAsync);
-            var tagHelperOutput = new TagHelperOutput(executionContext.TagName, executionContext.HTMLAttributes);
+            var tagHelperOutput = new TagHelperOutput(executionContext.TagName, executionContext.HTMLAttributes)
+            {
+                SelfClosing = executionContext.SelfClosing,
```

Comment:
since a `TagHelperOutput` is never constructed without immediately setting `SelfClosing`, the value should be a parameter to the constructor.  that way we know for sure nothing is creating these things without providing that setting.